rollup/internal/controller/permissionless_batches/minimal_recovery.go

```diff
@@ -0,0 +1,362 @@
+		return fmt.Errorf("bundle does not contain the correct batch: %d != %d", latestBundle.StartBatchIndex, latestBatch.Index)
+	}
+
+	log.Info("Bundle created", "index", latestBundle.Index, "hash", latestBundle.Hash, "StartBatchIndex", latestBundle.StartBatchIndex, "EndBatchIndex", latestBundle.EndBatchIndex, "starting block", firstChunkInBatch.StartBlockNumber, "ending block", lastChunkInBatch.EndBlockNumber)
+
+	return nil
+}
+
+// restoreMinimalPreviousState restores the minimal previous state required to be able to create new chunks, batches and bundles.
+func (r *MinimalRecovery) restoreMinimalPreviousState() (*orm.Chunk, *orm.Batch, *orm.Bundle, error) {
+	log.Info("Restoring previous state with", "L1 block height", r.cfg.RecoveryConfig.L1BlockHeight, "latest finalized batch", r.cfg.RecoveryConfig.LatestFinalizedBatch)
+
+	l1Client, err := ethclient.Dial(r.cfg.L1Config.Endpoint)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to connect to L1 client: %w", err)
+	}
+	reader, err := l1.NewReader(r.ctx, l1.Config{
+		ScrollChainAddress:    r.genesis.Config.Scroll.L1Config.ScrollChainAddress,
+		L1MessageQueueAddress: r.genesis.Config.Scroll.L1Config.L1MessageQueueAddress,
+	}, l1Client)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to create L1 reader: %w", err)
+	}
+
+	// 1. Sanity check user input: Make sure that the user's L1 block height is not higher than the latest finalized block number.
+	latestFinalizedL1Block, err := reader.GetLatestFinalizedBlockNumber()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get latest finalized L1 block number: %w", err)
+	}
+	if r.cfg.RecoveryConfig.L1BlockHeight > latestFinalizedL1Block {
+		return nil, nil, nil, fmt.Errorf("specified L1 block height is higher than the latest finalized block number: %d > %d", r.cfg.RecoveryConfig.L1BlockHeight, latestFinalizedL1Block)
+	}
+
+	log.Info("Latest finalized L1 block number", "latest finalized L1 block", latestFinalizedL1Block)
+
+	// 2. Make sure that the specified batch is indeed finalized on the L1 rollup contract and is the latest finalized batch.
+	var latestFinalizedBatch uint64
+	if r.cfg.RecoveryConfig.ForceLatestFinalizedBatch {
+		latestFinalizedBatch = r.cfg.RecoveryConfig.LatestFinalizedBatch
+	} else {
+		latestFinalizedBatch, err = reader.LatestFinalizedBatch(latestFinalizedL1Block)
+		if r.cfg.RecoveryConfig.LatestFinalizedBatch != latestFinalizedBatch {
+			return nil, nil, nil, fmt.Errorf("batch %d is not the latest finalized batch: %d", r.cfg.RecoveryConfig.LatestFinalizedBatch, latestFinalizedBatch)
+		}
+	}
+
+	var batchCommitEvent *l1.CommitBatchEvent
+	err = reader.FetchRollupEventsInRangeWithCallback(r.cfg.RecoveryConfig.L1BlockHeight, latestFinalizedL1Block, func(event l1.RollupEvent) bool {
+		if event.Type() == l1.CommitEventType && event.BatchIndex().Uint64() == r.cfg.RecoveryConfig.LatestFinalizedBatch {
+			batchCommitEvent = event.(*l1.CommitBatchEvent)
+			return false
```

Comment:
Added a comment for this. The callback function basically tells the `FetchRollupEventsInRangeWithCallback` function whether to continue. If true it will continue querying for rollup events, if false not. Since we're here only interested in finding the commit event of the `latestFinalizedBatch` we want to exit the search as soon as we found it.

Reply:
makes sense.